File: [143]重排链表.cpp
#include <cstdlib>
#include <stack>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	void reorderList(ListNode* head) {
		if (head == NULL)
			return;

		ListNode* p2 = head;
		for (ListNode* fast = head; fast && fast->next; p2 = p2->next, fast = fast->next->next);
		ListNode* p1 = head;
		stack<ListNode*> s;
		for (ListNode* p = p2; p != NULL; p = p->next)
			s.push(p);
		ListNode* p = head;
		while (p != p2) {
			ListNode* tmp = p->next;
			p->next = s.top();
			s.pop();
			p->next->next = tmp;
			p = tmp;
		}
		p->next = NULL;
	}
};

File: [143]重排链表.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # search for middle node and divide List into l1, l2
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        middle = slow
        l1 = head
        l2 = middle.next
        middle.next = None

        # reverse from middle to end
        prev = None
        cur = middle
        while prev:
            tmp_cur = cur.next
            cur.next = prev
            prev = cur
            cur = tmp_cur

        # merge l1, l2
        while l2:
            l1_tmp, l2_tmp = l1.next, l2.next

            l1.next = l2
            l1 = l1_tmp
            l2.next = l1
            l2 = l2_tmp
